File: src/lattice-kernel/lattice_kernel.m
%-----------------------------------------------------------------------------%
% SPDX-License-Identifier: GPL-3.0-or-later OR Apache-2.0
% CLONE_GATE: lattice_kernel_mercury
%
% lattice_kernel.m — Mercury production library
% Deterministic functions + multi-moded relations for the LATTICE kernel:
% lists, Peano, bit full-adder, SMT-stub SAT, path, FSM, interval refine.
% Build: mmc --make lattice_kernel   (needs a main module to link)
%   or compile as a library: mmc --make liblattice_kernel
%-----------------------------------------------------------------------------%

:- module lattice_kernel.
:- interface.

:- import_module bool.
:- import_module list.

% ---- types ----
:- type peano
    --->    z
    ;       s(peano).

:- type bit
    --->    b0
    ;       b1.

:- type smt
    --->    atom(string)
    ;       smt_not(smt)
    ;       smt_and(smt, smt).

:- type node == int.

:- type fsm_state
    --->    s0
    ;       s1.

:- type abs_val
    --->    conc(int)
    ;       top
    ;       interval(int, int).

% ---- lists (relational, reversible) ----
:- pred appendo(list(T), list(T), list(T)).
:- mode appendo(in, in, out) is det.
:- mode appendo(out, out, in) is multi.
:- mode appendo(in, out, in) is semidet.
:- mode appendo(out, in, in) is semidet.

:- pred membero(T, list(T)).
:- mode membero(in, in) is semidet.
:- mode membero(out, in) is nondet.

:- pred rembero(T, list(T), list(T)).
:- mode rembero(in, in, out) is semidet.
:- mode rembero(in, out, in) is nondet.

:- pred reverseo(list(T), list(T)).
:- mode reverseo(in, out) is det.
:- mode reverseo(out, in) is det.

:- pred lengtho(list(T), peano).
:- mode lengtho(in, out) is det.

:- pred distincto(list(T)).
:- mode distincto(in) is semidet.

% ---- Peano ----
:- pred nato(peano).
:- mode nato(in) is det.

:- pred addo(peano, peano, peano).
:- mode addo(in, in, out) is det.
:- mode addo(in, out, in) is semidet.
:- mode addo(out, in, in) is semidet.
:- mode addo(out, out, in) is multi.

:- pred mulo(peano, peano, peano).
:- mode mulo(in, in, out) is det.

:- func add_f(peano, peano) = peano.
:- func mul_f(peano, peano) = peano.

% ---- bits / adder ----
:- pred bit_xoro(bit, bit, bit).
:- mode bit_xoro(in, in, out) is det.
:- mode bit_xoro(in, out, in) is semidet.
:- mode bit_xoro(out, in, in) is semidet.
:- mode bit_xoro(in, in, in) is semidet.
:- mode bit_xoro(out, out, in) is multi.

:- pred bit_ando(bit, bit, bit).
:- mode bit_ando(in, in, out) is det.
:- mode bit_ando(in, in, in) is semidet.

:- pred half_addero(bit, bit, bit, bit).
:- mode half_addero(in, in, out, out) is det.

:- pred full_addero(bit, bit, bit, bit, bit).
:- mode full_addero(in, in, in, out, out) is det.
:- mode full_addero(in, in, in, in, in) is semidet.

:- pred ripple_addero(list(bit), list(bit), bit, list(bit), bit).
:- mode ripple_addero(in, in, in, out, out) is semidet.

% ---- SMT stub + path ----
:- pred smt_sato(smt, bool).
:- mode smt_sato(in, out) is det.
:- mode smt_sato(in, in) is semidet.

:- pred edgeo(node, node, smt).
:- mode edgeo(in, in, out) is nondet.
:- mode edgeo(out, out, out) is multi.
:- mode edgeo(in, out, out) is nondet.

:- pred patho(node, node, list(node), smt).
:- mode patho(in, in, out, out) is nondet.
:- mode patho(out, out, out, out) is nondet.

% ---- FSM / refine ----
:- pred fsmo(fsm_state, bit, fsm_state, bit).
:- mode fsmo(in, in, out, out) is det.
:- mode fsmo(in, in, in, in) is semidet.
:- mode fsmo(out, out, out, out) is multi.

:- pred refineo(abs_val, int).
:- mode refineo(in, in) is semidet.
:- mode refineo(in, out) is nondet.

:- implementation.

:- import_module int.
:- import_module string.

% ---- lists ----
appendo([], S, S).
appendo([A | D], S, [A | Res]) :-
    appendo(D, S, Res).

membero(X, [X | _]).
membero(X, [_ | D]) :-
    membero(X, D).

rembero(X, [X | D], D).
rembero(X, [A | D], [A | Res]) :-
    A \= X,
    rembero(X, D, Res).

reverseo(L, O) :-
    reverseo_acc(L, [], O).

:- pred reverseo_acc(list(T), list(T), list(T)).
:- mode reverseo_acc(in, in, out) is det.
:- mode reverseo_acc(out, in, in) is det.
reverseo_acc([], Acc, Acc).
reverseo_acc([A | D], Acc, O) :-
    reverseo_acc(D, [A | Acc], O).

lengtho([], z).
lengtho([_ | D], s(N)) :-
    lengtho(D, N).

distincto([]).
distincto([A | D]) :-
    not membero(A, D),
    distincto(D).

% ---- Peano ----
nato(z).
nato(s(P)) :-
    nato(P).

addo(z, Y, Y).
addo(s(X), Y, s(Z)) :-
    addo(X, Y, Z).

mulo(z, _, z).
mulo(s(X), Y, Z) :-
    mulo(X, Y, Z1),
    addo(Y, Z1, Z).

add_f(z, Y) = Y.
add_f(s(X), Y) = s(add_f(X, Y)).

mul_f(z, _) = z.
mul_f(s(X), Y) = add_f(Y, mul_f(X, Y)).

% ---- bits ----
bit_xoro(b0, b0, b0).
bit_xoro(b0, b1, b1).
bit_xoro(b1, b0, b1).
bit_xoro(b1, b1, b0).

bit_ando(b0, b0, b0).
bit_ando(b0, b1, b0).
bit_ando(b1, b0, b0).
bit_ando(b1, b1, b1).

half_addero(X, Y, R, C) :-
    bit_xoro(X, Y, R),
    bit_ando(X, Y, C).

full_addero(Cin, X, Y, R, Cout) :-
    half_addero(X, Y, W, C1),
    half_addero(W, Cin, R, C2),
    bit_xoro(C1, C2, Cout).

ripple_addero([], [], Cin, [], Cin).
ripple_addero([X | Xs], [Y | Ys], Cin, [R | Rs], Cout) :-
    full_addero(Cin, X, Y, R, C1),
    ripple_addero(Xs, Ys, C1, Rs, Cout).

% ---- SMT stub: reject phi /\ ~phi at top-level flatten ----
smt_sato(Phi, Sat) :-
    atoms(Phi, As),
    ( unsat_pair(As) -> Sat = no ; Sat = yes ).

:- pred atoms(smt::in, list(smt)::out) is det.
atoms(smt_and(A, B), Xs) :-
    atoms(A, Xs1),
    atoms(B, Xs2),
    appendo(Xs1, Xs2, Xs).
atoms(smt_not(P), [smt_not(P)]).
atoms(atom(S), [atom(S)]).

:- pred unsat_pair(list(smt)::in) is semidet.
unsat_pair(As) :-
    membero(P, As),
    membero(smt_not(P), As).

edgeo(0, 1, atom("bv_slt_x_y")).
edgeo(1, 2, atom("bv_slt_y_z")).
edgeo(2, 3, atom("bv_sgt_z_x")).
edgeo(3, 1, smt_and(atom("p"), smt_not(atom("q")))).

patho(X, Y, [Y], Phi) :-
    edgeo(X, Y, Phi),
    smt_sato(Phi, yes).
patho(X, Y, [Z | P2], smt_and(E, Phi2)) :-
    edgeo(X, Z, E),
    patho(Z, Y, P2, Phi2),
    not membero(Z, P2),
    smt_sato(smt_and(E, Phi2), yes).

% ---- FSM / refine ----
fsmo(s0, b0, s0, b0).
fsmo(s0, b1, s1, b0).
fsmo(s1, b0, s0, b1).
fsmo(s1, b1, s1, b1).

refineo(conc(C), C).
refineo(top, _).
refineo(interval(Lo, Hi), N) :-
    N >= Lo,
    N =< Hi.

:- end_module lattice_kernel.

File: src/lattice-kernel/lattice_kernel.pl
%% SPDX-License-Identifier: GPL-3.0-or-later OR Apache-2.0
%% CLONE_GATE: lattice_kernel_pl
%%
%% lattice_kernel.pl — Relational Lattice Kernel (ISO Prolog)
%%
%% Same relations as mu_kanren.scm as portable ISO predicates.
%% Optional CLP(FD) adder via use_module(library(clpfd)).
%% Compatible with SWI-Prolog, GNU Prolog, SICStus.
%%
%% Run: swipl -s lattice_kernel.pl
%%      ?- demo.

:- use_module(library(lists)).

%% ============================================================
%% 1. LIST RELATIONS
%% ============================================================

conso(A, D, [A|D]).
nullo([]).
caro([A|_], A).
cdro([_|D], D).
pairo([_|_]).

listo([]).
listo([_|D]) :- listo(D).

appendo([], S, S).
appendo([A|D], S, [A|R]) :- appendo(D, S, R).

membero(X, [X|_]).
membero(X, [_|T]) :- membero(X, T).

rembero(_, [], []).
rembero(X, [X|D], D) :- !.
rembero(X, [A|D], [A|R]) :- X \= A, rembero(X, D, R).

reverseo(L, O) :- rev_acco(L, [], O).
rev_acco([], Acc, Acc).
rev_acco([A|D], Acc, O) :- rev_acco(D, [A|Acc], O).

lengtho([], z).
lengtho([_|D], s(N)) :- lengtho(D, N).

distincto([]).
distincto([A|D]) :- \+ member(A, D), distincto(D).

permuteo([], []).
permuteo([A|D], O) :-
    permuteo(D, R),
    inserto(A, R, O).

inserto(X, L, [X|L]).
inserto(X, [A|D], [A|R]) :- inserto(X, D, R).

zipo([], [], []).
zipo([A|DA], [B|DB], [[A,B]|R]) :- zipo(DA, DB, R).

takeo(0, _, []) :- !.
takeo(s(N), [A|D], [A|R]) :- takeo(N, D, R).

dropo(0, L, L) :- !.
dropo(s(N), [_|D], O) :- dropo(N, D, O).

filtero(_, [], []).
filtero(P, [A|D], Out) :-
    ( call(P, A) ->
        filtero(P, D, R), Out = [A|R]
    ; filtero(P, D, Out)
    ).

%% ============================================================
%% 2. PEANO ARITHMETIC
%% ============================================================

zeroo(z).
succo(N, s(N)).
nato(z).
nato(s(N)) :- nato(N).

addo(z, Y, Y).
addo(s(X), Y, s(Z)) :- addo(X, Y, Z).

mulo(z, _, z).
mulo(s(X), Y, Z) :-
    mulo(X, Y, Z1),
    addo(Y, Z1, Z).

leq_nato(z, _).
leq_nato(s(X), s(Y)) :- leq_nato(X, Y).

%% ============================================================
%% 3. BIT ARITHMETIC — FULL ADDER
%% ============================================================

bito(0).
bito(1).

half_addo(0, 0, 0, 0).
half_addo(0, 1, 1, 0).
half_addo(1, 0, 1, 0).
half_addo(1, 1, 0, 1).

full_addo(A, B, Cin, S, Cout) :-
    bito(A), bito(B), bito(Cin),
    half_addo(A, B, W, C1),
    half_addo(W, Cin, S, C2),
    ( C1 =:= 1 ; C2 =:= 1 -> Cout = 1 ; Cout = 0 ).

%% CLP(FD) adder (loaded conditionally)
:- ( catch(use_module(library(clpfd)), _, true) ->
       assert(clpfd_available)
   ; true
   ).

full_addo_clp(A, B, Cin, S, Cout) :-
    clpfd_available, !,
    [A, B, Cin, S, Cout] ins 0..1,
    S #= (A + B + Cin) mod 2,
    Cout #= (A + B + Cin) // 2.
full_addo_clp(A, B, Cin, S, Cout) :-
    full_addo(A, B, Cin, S, Cout).

%% n-bit ripple-carry adder
addero([], [], [], Cin, Cin).
addero([A|As], [B|Bs], [S|Ss], Cin, Cout) :-
    full_addo(A, B, Cin, S, Cmid),
    addero(As, Bs, Ss, Cmid, Cout).

%% ============================================================
%% 4. RELATIONAL INTERPRETER (evalo)
%% ============================================================

:- dynamic edb/3, edb/4, fsm_transition/5.

lookupo(X, [[X,V]|_], V) :- !.
lookupo(X, [[Y,_]|Rest], V) :-
    X \= Y,
    lookupo(X, Rest, V).

eval_expro(quote(Q), _, Q).
eval_expro(X, Env, V) :-
    atom(X), \+ X = quote, \+ X = lambda, \+ X = cons,
    \+ X = car, \+ X = cdr, \+ X = if, \+ X = 'null?',
    lookupo(X, Env, V).
eval_expro(lambda(P, B), Env, closure(P, B, Env)).
eval_expro(app(Rator, Rand), Env, V) :-
    eval_expro(Rator, Env, Clo),
    eval_expro(Rand, Env, Arg),
    applyo(Clo, Arg, V).
eval_expro(cons(A, D), Env, [VA|VD]) :-
    eval_expro(A, Env, VA),
    eval_expro(D, Env, VD).
eval_expro(car(E), Env, V) :-
    eval_expro(E, Env, [V|_]).
eval_expro(cdr(E), Env, V) :-
    eval_expro(E, Env, [_|V]).
eval_expro(if(C, T, El), Env, V) :-
    eval_expro(C, Env, CV),
    ( CV \= false -> eval_expro(T, Env, V)
    ; eval_expro(El, Env, V)
    ).

applyo(closure(P, B, Env), Arg, V) :-
    eval_expro(B, [[P,Arg]|Env], V).

evalo(Exp, Val) :- eval_expro(Exp, [], Val).

%% ============================================================
%% 5. SMT STUB
%% ============================================================

smt_and(A, B, and(A, B)).
smt_not(A, not(A)).
smt_var(N, var(N)).

is_sat(Phi) :-
    collect_atoms(Phi, [], Atoms),
    \+ contradiction(Atoms).

collect_atoms(and(A, B), Acc, Out) :- !,
    collect_atoms(A, Acc, Acc1),
    collect_atoms(B, Acc1, Out).
collect_atoms(X, Acc, [X|Acc]).

contradiction(Atoms) :-
    member(not(P), Atoms),
    member(P, Atoms).

%% ============================================================
%% 6. EDB FACT STORE
%% ============================================================

assert_edb(Rel, A, B) :- assertz(edb(Rel, A, B)).
assert_edb(Rel, A, B, C) :- assertz(edb(Rel, A, B, C)).

factso(Rel, A, B) :- edb(Rel, A, B).
factso(Rel, A, B, C) :- edb(Rel, A, B, C).

%% ============================================================
%% 7. PARAMETERIZED PATH / REACHABILITY
%% ============================================================

%% patho(+Rel, +X, ?Y, -Path, -Phi)
patho(Rel, X, Y, path(Y), Phi) :-
    factso(Rel, X, Y, Phi),
    is_sat(Phi).
patho(Rel, X, Y, path(Z, P), and(E, Phi2)) :-
    factso(Rel, X, Z, E),
    is_sat(E),
    patho(Rel, Z, Y, P, Phi2).

reacho(_, S, S).
reacho(Rel, S, T) :-
    edb(Rel, S, M),
    reacho(Rel, M, T).

%% ============================================================
%% 8. FSM RELATIONS
%% ============================================================

register_fsm(Id) :-
    assertz(fsm_transition(Id, s0, 0, s0, 0)),
    assertz(fsm_transition(Id, s0, 1, s1, 0)),
    assertz(fsm_transition(Id, s1, 0, s0, 1)),
    assertz(fsm_transition(Id, s1, 1, s1, 1)).

fsmo(Id, St, Inp, Nst, Out) :-
    fsm_transition(Id, St, Inp, Nst, Out).

%% fsm_traceo(+Id, +S0, +Inputs, -Outputs)
fsm_traceo(_, _, [], []).
fsm_traceo(Id, S, [I|Rest], [O|RestO]) :-
    fsmo(Id, S, I, S1, O),
    fsm_traceo(Id, S1, Rest, RestO).

%% ============================================================
%% 9. REFINEMENT LATTICE
%% ============================================================

refineo(conc(C), C, refined).
refineo(top, _, refined).
refineo(interval(Lo, Hi), N, refined) :-
    leq_nato(Lo, N),
    leq_nato(N, Hi).

refineo_nat(conc(C), C).
refineo_nat(top, _).
refineo_nat(interval(Lo, Hi), N) :-
    integer(Lo), integer(Hi), integer(N),
    Lo =< N, N =< Hi.

%% ============================================================
%% 10. HIGHER-ORDER LIST OPERATIONS
%% ============================================================

mapo(_, [], []).
mapo(Goal, [A|D], [B|R]) :-
    call(Goal, A, B),
    mapo(Goal, D, R).

foldo(_, [], Acc, Acc).
foldo(Goal, [A|D], Acc, Out) :-
    call(Goal, A, Acc, NAcc),
    foldo(Goal, D, NAcc, Out).

%% ============================================================
%% 11. SEED DATA
%% ============================================================

seed_db :-
    retractall(edb(_, _, _)),
    retractall(edb(_, _, _, _)),
    retractall(fsm_transition(_, _, _, _, _)),
    assert_edb(edge, 0, 1, bv_slt(var(x), var(y))),
    assert_edb(edge, 1, 2, bv_slt(var(y), var(z))),
    assert_edb(edge, 2, 3, and(var(p), not(var(q)))),
    assert_edb(edge, 0, 2),
    assert_edb(edge, 1, 3),
    register_fsm(fsm0).

%% ============================================================
%% 12. DEMO / SMOKE TEST
%% ============================================================

demo :-
    seed_db,
    format('=== lattice_kernel.pl ===~n'),

    format('~nappendo([1,2],[3],?): '),
    findall(Q, appendo([1,2],[3],Q), R1),
    format('~w~n', [R1]),

    format('addo(s(z),s(s(z)),?): '),
    findall(Q, addo(s(z),s(s(z)),Q), R2),
    format('~w~n', [R2]),

    format('full_addo(1,1,0,S,C): '),
    findall(S-C, full_addo(1,1,0,S,C), R3),
    format('~w~n', [R3]),

    format('evalo: '),
    findall(V, evalo(app(lambda(x,x),quote(hello)),V), R4),
    format('~w~n', [R4]),

    format('reacho from 0: '),
    findall(T, reacho(edge,0,T), R5),
    sort(R5, R5s),
    format('~w~n', [R5s]),

    format('fsm_traceo [0,1,0]: '),
    findall(O, fsm_traceo(fsm0,s0,[0,1,0],O), R6),
    format('~w~n', [R6]),

    format('refineo interval: '),
    findall(R, refineo(interval(s(z),s(s(s(z)))),s(s(z)),R), R7),
    format('~w~n', [R7]),

    format('is_sat true: '),
    ( is_sat(and(var(p), not(var(q)))) -> format('sat~n') ; format('unsat~n') ),
    format('is_sat contradiction: '),
    ( is_sat(and(var(p), not(var(p)))) -> format('sat~n') ; format('unsat~n') ),

    format('~n=== COMPLETE ===~n').

:- initialization(demo, main).
